Edited version of english translation (#11)
inem/make-handbook: modern-make-handbook-en.md

[![](https://i.imgur.com/a1MOR8T.png)](https://makefile.site)

[TOC]

<div style="page-break-after: always;"></div>
## 1. Making your library of shortcuts

– What cats say when they’re hungry?

– Meow!

– What dogs say when they smell danger?

– Woof-woof!

– And what does Bob say when he’s deploying project?

– `ansible-playbook -i inventory/production --tags “deploy” app-server.yml -vvv --become-user=app --extra-vars=extra.txt --vault-password-file="~/.ansible/vault.txt"`

This is what we’re going to fix.

It is not gonna be hard. It will be easy and pleasant process. Just like organizing a closet.

Open `Makefile` and put your mega-command in there:

```make
deploy:
	ansible-playbook -i inventory/production --tags “deploy” app-server.yml -vvv --become-user=app --extra-vars=extra.txt --vault-password-file="~/.ansible/vault.txt"
```

Next time Bob will have to say just `make deploy.

Cool, huh?  It means that

- Bob saves time because he doesn't have to remember all the details of the deploy command
- Bob will never make a mistake in the deploy command
- Bob is not going to freak out, when Ansible renames half of their flags

Let’s move on.

> Often, things that supposed to be simple, like `rails server`,  owergrow with addtional debilitating details: `bundle exec bin/rails server -p 3001 RAILS_ENV=development`. 

Reviewer: ```suggestion
Often, things that supposed to be simple, like `rails server`,  overgrow with addtional debilitating details: `bundle exec bin/rails server -p 3001 RAILS_ENV=development`. 
```

> Luckily we already know what to do::

Reviewer: ```suggestion
Luckily we already know what to do:
```

```make
deploy: test
	ansible-playbook -i inventory/production --tags ‘deploy’ # ...
```
This means that each time you run `make deploy`, `make test` is called first. And only if it succeeds, `deploy` will be executed.

<div style="page-break-after: always;"></div>
## 5. Aliases

Alice adds a new command to run migrations, but Bob keeps forgetting how it is called. He run `make dbmigrate`, `make db_migrate`, and even `make db:migrate`, but kept getting error: 

```
make: *** No rule to make target '*'.  Stop.
```

We can fix this problem with aliases! Unfortunately, the latter will not work, because `make` command names can't contain colons. But for the rest of the typos we can do it easily, even without copying and pasting:

```make
db-migrate:
	bundle exec bin/rails db:migrate

db_migrate: db-migrate
dbmigrate: db-migrate
```

Sometimes it is hard to come up with handy name for a shortcut. In this case just create a bunch of aliases and keep the most used one after a while.

<div style="page-break-after: always;"></div>
## 6. Multiline commands

After a while, the team decided to replace `make db-migrate` with just `make db`, which is impossible to forget.

> Of course, they added `db` to `.PHONY:`, because Rails projects have `db`directory as well.

Reviewer: ```suggestion
Of course, they added `db` to `.PHONY:`, because Rails projects have `db` directory as well.
```

	scp app@staging-server.dev:/path/to/app/$(F)/ ./
```

And we call it like this:

`make staging-fetch F=db/dump.tgz`

<div style="page-break-after: always;"></div>
## 11. Seamless arguments

What if we could simplify the command by skipping the name of the argument, since we have only one here?

Could we just do `make staging-fetch db/dump.tgz`?

The answer is yes! We can do it with a bit of black magic. We have to add the following statement into the `Makefile`:

```make
ARGS = $(filter-out $@,$(MAKECMDGOALS))
%:
	@:
```

Your shortcut will turn into something that looks like this:

```
staging-fetch:
	scp app@staging-server.dev:/path/to/app/$(ARGS)/ ./
```

We called it "balck magic" for a reason. This trick has one annoying side effect. After the script is executed, you get an error message like this:

```
make: *** No rule to make target ‘db/dump.tgz’. Stop.
```

Here is a [post on StackOverflow](https://stackoverflow.com/a/6273809/1334666) that explains how it works and even contains a fix for this problem. Although, Bob couldn‘t make it work, but he decided that this is the price he is ready to pay.

<div style="page-break-after: always;"></div>
## 12. Advanced scripting

> Another day DevOps team announces that for now on each feature branch will be deployed to a separate staging server.

Reviewer: ```suggestion
Another day the DevOps team announces that from now on each feature branch will be deployed to a separate staging server.
```
